libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,555 @@
+const QMap<int, int>& RealizedHarmony::notes() const
+      {
+      Q_ASSERT(!_dirty);
+      //with the way that the code is currently structured, there should be no way to
+      //get to this function with dirty flag set although in the future it may be
+      //better to just update if dirty here
+      return _notes;
+      }
+
+//---------------------------------------------------
+//   notes
+///   returns the list of notes
+//---------------------------------------------------
+const QMap<int, int> RealizedHarmony::generateNotes(int rootTpc, int bassTpc,
+                                                bool literal, Voicing voicing, int transposeOffset) const
+      {
+      QMap<int, int> notes;
+      int rootPitch = tpc2pitch(rootTpc) + transposeOffset;
+      //euclidian mod, we need to treat this new pitch as a pitch between
+      //0 and 11, so that voicing remains consistent across transposition
+      if (rootPitch < 0)
+            rootPitch += PITCH_DELTA_OCTAVE;
+      else
+            rootPitch %= PITCH_DELTA_OCTAVE;
+
+      //create root note or bass note in second octave below middle C
+      if (bassTpc != Tpc::TPC_INVALID && voicing != Voicing::ROOT_ONLY)
+            notes.insert(tpc2pitch(bassTpc) + transposeOffset
+                        + 3*PITCH_DELTA_OCTAVE, bassTpc);
+      else
+            notes.insert(rootPitch + 3*PITCH_DELTA_OCTAVE, rootTpc);
+
+
+      switch (voicing) {
+            case Voicing::ROOT_ONLY: //already added root/bass so we are good
+                  break;
+            case Voicing::AUTO: //auto is close voicing for now since it is the most robust
+            case Voicing::CLOSE://Voices notes in close position in the first octave above middle C
+                  // FALLTHROUGH
+                  {
+                  notes.insert(rootPitch + 5*PITCH_DELTA_OCTAVE, rootTpc);
+                  //ensure that notes fall under a specific range
+                  //for now this range is between 5*12 and 6*12
+                  QMap<int, int> intervals = getIntervals(rootTpc, literal);
+                  QMapIterator<int, int> i(intervals);
+                  while (i.hasNext()) {
+                        i.next();
+                        notes.insert((rootPitch + (i.key() % 128)) % PITCH_DELTA_OCTAVE +
+                                      5*PITCH_DELTA_OCTAVE, i.value());
+                        }
+                  }
+                  break;
+            case Voicing::DROP_2:
+                  {
+                  //select 4 notes from list
+                  QMap<int, int> intervals = normalizeNoteMap(getIntervals(rootTpc, literal), rootTpc, rootPitch, 4);
+                  QMapIterator<int, int> i(intervals);
+                  i.toBack();
+
+                  int counter = 0; //counter to drop the second note
+                  while (i.hasPrevious()) {
+                        i.previous();
+                        if (++counter == 2)
```

Comment:
I'd suggest defining the default octave as a constant, then using either that or that-1 here, rather than hardcoding both the 4 & 5 here.